src/workerd/jsg/jsg.h

```diff
@@ -1326,17 +1326,44 @@ using PromiseForResult = Promise<RemovePromise<ReturnType<Func, Param, passLock>
 
 class ModuleRegistry;
 
+class ContextGlobal {
+  // All jsg resources that are used as context globals must eventually extend this type.
+  // The lifecycle of this object is tied to a v8 context.
+
+public:
+  ContextGlobal() {}
+
+  KJ_DISALLOW_COPY_AND_MOVE(ContextGlobal);
+
+  ModuleRegistry& getModuleRegistry() { return *moduleRegistry; }
+
+private:
+  kj::Own<ModuleRegistry> moduleRegistry;
+
+  void setModuleRegistry(kj::Own<ModuleRegistry> registry) {
+    moduleRegistry = kj::mv(registry);
+  }
+
+  template <typename, typename>
+  friend class ResourceWrapper;
+};
+
+class ContextGlobalObject: public Object, public ContextGlobal { };
+// Convenient shortcut
```

Comment:
This could probably use a more lengthy comment.

> All global objects must inherit both `jsg::Object` and `jsg::ContextGlobal`. In most cases, this can be achieved by inheriting this class. However, sometimes a global object itself needs to extend some other resource type, where that other type is not always used as a context global object. For example, `WorkerGlobalScope` inherits `EventTarget`. `EventTarget` itself inherits `jsg::Object`, so if `WorkerGlobalScope` tried to inherit both `EventTarget` and `jsg::ContextGlobalObject`, it would end up inheriting multiple copies of `jsg::Object` (which won't work!). Instead, it needs to inherit `ContextGlobal` directly. Less-complicated contexts that don't need to extend another type can inherit `ContextGlobalObject`.

... though to be honest, I think this explanation is more complicated than the "convenient shortcut" is worth. I would be more inclined to just delete this type and just require all contexts multiply-inherit `jsg::Object` and `jsg::ContextGlobal`.

Reply:
I have removed ContextGlobalObject definition and used it only in tests with a lot of contexts. It is one liner, so it wouldn't hurt.